src/Mvc/Mvc.Core/src/Infrastructure/ActionSelectionTable.cs

```diff
@@ -0,0 +1,155 @@
+using System.Globalization;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.Infrastructure
+{
+    // Common infrastructure for things that look up actions by route values.
+    //
+    // The ActionSelectionTablee stores a mapping of route-values -> items for each known set of
+    // of route-values. We actually build two of these mappings, one for case-sensitive (fast path) and one for
+    // case-insensitive (slow path).
+    //
+    // This is necessary because MVC routing/action-selection is always case-insensitive. So we're going to build
+    // a case-sensitive dictionary that will behave like the a case-insensitive dictionary when you hit one of the
+    // canonical entries. When you don't hit a case-sensitive match it will try the case-insensitive dictionary
+    // so you still get correct behaviors.
+    //
+    // The difference here is because while MVC is case-insensitive, doing a case-sensitive comparison is much
+    // faster. We also expect that most of the URLs we process are canonically-cased because they were generated
+    // by Url.Action or another routing api.
+    //
+    // This means that for a set of actions like:
+    //      { controller = "Home", action = "Index" } -> HomeController::Index1()
+    //      { controller = "Home", action = "index" } -> HomeController::Index2()
+    //
+    // Both of these actions match "Index" case-insensitively, but there exist two known canonical casings,
+    // so we will create an entry for "Index" and an entry for "index". Both of these entries match **both**
+    // actions.
+    internal class ActionSelectionTable<TItem>
+    {
+        private ActionSelectionTable(
+            int version, 
+            string[] routeKeys,
+            Dictionary<string[], List<TItem>> ordinalEntries,
+            Dictionary<string[], List<TItem>> ordinalIgnoreCaseEntries)
+        {
+            Version = version;
+            RouteKeys = routeKeys;
+            OrdinalEntries = ordinalEntries;
+            OrdinalIgnoreCaseEntries = ordinalIgnoreCaseEntries;
+        }
+        
+        public int Version { get; }
+
+        public string[] RouteKeys { get; }
+
+        public Dictionary<string[], List<TItem>> OrdinalEntries { get; }
+
+        public Dictionary<string[], List<TItem>> OrdinalIgnoreCaseEntries { get; }
+
+        public static ActionSelectionTable<ActionDescriptor> Create(ActionDescriptorCollection actions)
+        {
+            return CreateCore<ActionDescriptor>(
+
+                // We need to store the version so the cache can be invalidated if the actions change.
+                actions.Version,
+
+                // For action selection, ignore attribute routed actions
+                actions.Items.Where(a => a.AttributeRouteInfo == null),
+
+                a => a.RouteValues.Keys,
+                (a, key) =>
```

Comment:
nit: use named parameters with these args to make it easier to understand what each is for